flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+Before starting to implement anything, it is best to submit this interface for review. Having the interface
+already submitted can also allow to implement the lowering part in parallel of the runtime implementation.
+See `lib/runtime/character.h` for examples of runtime interfaces.
+
+The design must decide if one or more functions should form the runtime interface.
+It may make sense to have simple entry points for easy cases, and a more complex entry points
+with descriptors to cover all the other cases. When in doubt, it is best to start with a single
+entry point that cover all cases, and leave entry point specialization for later, based on a measurable
+performance gain on actual programs.
+
+The implementation is done in C++, but there is a hard requirement in not depending on
+the C++ runtime library. That is because it is not wished that Fortran runtime depends
+upon libstdc++ or other C++ runtime libraries. This means that templates and classes
+are OK, `#include <c[library]>` should be OK, but any other includes should be avoided
+unless proven they do not bring C++ runtime in.
+
+The implementation is done in a related `.cpp` file in the same folder. See
+`lib/runtime/character.cpp` for an example.
+
+#### Tests
+To be defined. Ideally, we would want end-to-end tests from Fortran, but it is
+not yet clear where to put them in llvm. Unit tests can also be defined. See
+`unittests/Runtime/character.cpp` for an example.
+
+### Lowering
+- See: `lib/Lower/IntrinsicCall.cpp`
+- Where to submit patch: fir-dev branch of https://github.com/flang-compiler/f18-llvm-project.
+
+#### Implementation
+Lowering of intrinsic is driven by the `IntrinsicLibrary` class. Its member function
+`gen{intrinsic name}` lowers "intrinsic name". These functions have standardized interfaces.
+There is a table in this file that maps an intrinsic name to the function that lowers it.
+To add support for an intrinsic, a new member function containing the lowering implementation
+must be created and mapped into the table.
+
+The lowering implementation can be inlined if it can easily be implemented in a few fir or
+standard mlir dialect operations. Otherwise, the lowering implementation should generate code
+that emits a call to the runtime implementing the intrinsic.
+
+There are three standardized interfaces to choose from for implementation member functions:
+- `mlir::Value genXXX(mlir::Type, llvm::ArrayRef<mlir::Value>)`
+- `fir::ExtendedValue genXXX(mlir::Type, llvm::ArrayRef<fir::ExtendedValue>)`
+- `void genXXX(llvm::ArrayRef<fir::ExtendedValue>)`
+
+The first interface is to be used for elemental numerical and logical intrinsic functions,
+see `genAbs` for an example.
+The second interface is to be used for other intrinsic functions, see `genTrim` for an example.
+The third interface is to be used for intrinsic subroutine, see `genDateAndTime` for an example.
+
+It is worth being familiar with `fir::ExtendedValue`. They described an entity (symbol,
```

Comment:
It is worthwhile becoming familiar ...
s/They/It/ (As in it is a set of hammers.)

Reply:
Have to stop here for now...